src/Symfony/Component/Messenger/Transport/MemoryTransport.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+class MemoryTransport implements TransportInterface
+{
+    private $envelopes = array();
+    private $bus;
+
+    public function __construct(MessageBusInterface $bus)
+    {
+        $this->bus = $bus;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(callable $handler): void
+    {
+        while (\count($this->envelopes) > 0) {
+            $handler(array_shift($this->envelopes));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stop(): void
+    {
+        $this->envelopes = array();
```

Comment:
Instead of emptying, we may just use a `$stopped` flag used in receive to skip handling as in https://github.com/symfony/symfony/blob/52f0a2c3053f3381a1be1843d125801bfd537845/src/Symfony/Component/Messenger/Transport/NullTransport.php#L40 so we keep envelopes list on exception, but don't re-dispatch them.